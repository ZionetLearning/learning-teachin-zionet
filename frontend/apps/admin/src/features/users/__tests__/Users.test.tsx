import {
  QueryClient,
  QueryClientProvider,
  type UseQueryResult,
} from "@tanstack/react-query";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { Users } from "../index";
import { User } from "@app-providers";

vi.mock("react-i18next", async () => {
  const actual =
    await vi.importActual<typeof import("react-i18next")>("react-i18next");
  return {
    ...actual,
    useTranslation: () => ({
      t: (k: string) => k,
      i18n: { changeLanguage: vi.fn() },
    }),
  };
});

vi.mock("react-toastify", () => ({
  toast: { success: vi.fn(), error: vi.fn() },
}));

vi.mock("@admin/api", () => ({
  useGetAllUsers: vi.fn(),
  useUpdateUserByUserId: vi.fn(),
  useDeleteUserByUserId: vi.fn(),
}));

vi.mock("@app-providers", async () => {
  const actual = (await vi.importActual("@app-providers")) as Record<
    string,
    unknown
  >;
  return {
    ...actual,
    useCreateUser: vi.fn(),
  };
});

const { useGetAllUsers, useUpdateUserByUserId, useDeleteUserByUserId } =
  vi.mocked(await import("@admin/api")) as unknown as {
    useGetAllUsers: ReturnType<typeof vi.fn>;
    useUpdateUserByUserId: ReturnType<typeof vi.fn>;
    useDeleteUserByUserId: ReturnType<typeof vi.fn>;
  };

const { useCreateUser } = vi.mocked(
  await import("@app-providers"),
) as unknown as {
  useCreateUser: ReturnType<typeof vi.fn>;
};

const rq = (
  over: Partial<UseQueryResult<User[], Error>>,
): UseQueryResult<User[], Error> =>
  ({
    data: undefined,
    error: null,
    isLoading: false,
    isError: !!over.error,
    isSuccess: !!over.data && !over.error,
    fetchStatus: "idle",
    status: over.isLoading ? "pending" : over.error ? "error" : "success",
    refetch: vi.fn(),
    failureCount: 0,
    isPending: over.isLoading ?? false,
    isFetched: true,
    isInitialLoading: false,
    isFetching: false,
    dataUpdatedAt: Date.now(),
    errorUpdatedAt: 0,
    isPaused: false,
    ...over,
  }) as UseQueryResult<User[], Error>;

const sampleUsers: User[] = [
  {
    userId: "u1",
    email: "alpha@example.com",
    firstName: "Alice",
    lastName: "Anderson",
  },
  {
    userId: "u2",
    email: "beta@example.com",
    firstName: "Bob",
    lastName: "Baker",
  },
];

let qc: QueryClient;
const renderUsers = () =>
  render(
    <QueryClientProvider client={qc}>
      <Users />
    </QueryClientProvider>,
  );

beforeEach(() => {
  vi.clearAllMocks();
  qc = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  if (globalThis.crypto && "randomUUID" in globalThis.crypto) {
    try {
      vi.spyOn(globalThis.crypto, "randomUUID").mockReturnValue(
        "123e4567-e89b-12d3-a456-426614174000",
      );
    } catch {
      const existing = globalThis.crypto as Crypto;
      Object.defineProperty(globalThis, "crypto", {
        value: {
          ...existing,
          randomUUID: () => "123e4567-e89b-12d3-a456-426614174000",
        },
        configurable: true,
      });
    }
  } else {
    Object.defineProperty(globalThis, "crypto", {
      value: { randomUUID: () => "123e4567-e89b-12d3-a456-426614174000" },
      configurable: true,
    });
  }
});

describe("<Users />", () => {
  it("matches snapshot (with users)", () => {
    useGetAllUsers.mockReturnValue(rq({ data: sampleUsers }));
    useCreateUser.mockReturnValue({ mutate: vi.fn(), isPending: false });
    useUpdateUserByUserId.mockImplementation(() => ({
      mutate: vi.fn(),
      isPending: false,
    }));
    useDeleteUserByUserId.mockImplementation(() => ({
      mutate: vi.fn(),
      isPending: false,
    }));

    const { asFragment } = renderUsers();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders loading state", () => {
    useGetAllUsers.mockReturnValue(rq({ isLoading: true }));
    useCreateUser.mockReturnValue({ mutate: vi.fn(), isPending: false });
    renderUsers();
    expect(screen.getByText("pages.users.loadingUsers")).toBeInTheDocument();
  });

  it("renders error state", () => {
    useGetAllUsers.mockReturnValue(
      rq({ error: new Error("Boom"), data: undefined }),
    );
    useCreateUser.mockReturnValue({ mutate: vi.fn(), isPending: false });
    renderUsers();
    expect(screen.getByText("pages.users.userNotFound")).toBeInTheDocument();
  });

  it("submits create user form", async () => {
    type Vars = {
      userId: string;
      email: string;
      password: string;
      firstName: string;
      lastName: string;
    };
    interface Handlers {
      onSuccess?: () => void;
      onError?: (e: Error) => void;
      onSettled?: () => void;
    }
    const mutate = vi.fn((_: Vars, opts?: Handlers) => {
      opts?.onSuccess?.();
      opts?.onSettled?.();
    });
    useGetAllUsers.mockReturnValue(rq({ data: sampleUsers }));
    useCreateUser.mockReturnValue({ mutate, isPending: false });
    useUpdateUserByUserId.mockImplementation(() => ({
      mutate: vi.fn(),
      isPending: false,
    }));
    useDeleteUserByUserId.mockImplementation(() => ({
      mutate: vi.fn(),
      isPending: false,
    }));

    renderUsers();
    fireEvent.change(screen.getByPlaceholderText(/user@example.com/i), {
      target: { value: "new@example.com" },
    });
    fireEvent.change(screen.getByPlaceholderText(/John/i), {
      target: { value: "NewFirst" },
    });
    fireEvent.change(screen.getByPlaceholderText(/Doe/i), {
      target: { value: "NewLast" },
    });
    fireEvent.change(screen.getByPlaceholderText(/\*\*\*\*\*\*/i), {
      target: { value: "Secret123!" },
    });
    fireEvent.click(screen.getByTestId("users-create-submit"));
    await waitFor(() => expect(mutate).toHaveBeenCalledTimes(1));
    const arg = mutate.mock.calls[0][0];
    expect(arg.email).toBe("new@example.com");
    expect(arg.password).toBe("Secret123!");
    expect(arg.firstName).toBe("NewFirst");
    expect(arg.lastName).toBe("NewLast");
    expect(arg.userId).toBe("123e4567-e89b-12d3-a456-426614174000");
  });

  it("updates a user via inline edit form (partial fields only)", () => {
    const updateMutate = vi.fn();
    useGetAllUsers.mockReturnValue(rq({ data: [sampleUsers[0]] }));
    useCreateUser.mockReturnValue({ mutate: vi.fn(), isPending: false });
    useUpdateUserByUserId.mockImplementation(() => ({
      mutate: updateMutate,
      isPending: false,
    }));
    useDeleteUserByUserId.mockImplementation(() => ({
      mutate: vi.fn(),
      isPending: false,
    }));

    renderUsers();
    fireEvent.click(screen.getByRole("button", { name: /update/i }));
    const emailInput = screen.getByPlaceholderText("email") as HTMLInputElement;
    fireEvent.change(emailInput, { target: { value: "changed@example.com" } });
    // Provide a first name change so payload includes it while omitting lastName
    const firstNameInput = screen.getByPlaceholderText(
      "first name",
    ) as HTMLInputElement;
    fireEvent.change(firstNameInput, { target: { value: "Charlie" } });
    fireEvent.click(screen.getByRole("button", { name: /save/i }));
    expect(updateMutate).toHaveBeenCalledTimes(1);
    expect(updateMutate.mock.calls[0][0]).toEqual({
      email: "changed@example.com",
      firstName: "Charlie",
    });
  });

  it("deletes a user", () => {
    const deleteMutate = vi.fn();
    useGetAllUsers.mockReturnValue(rq({ data: [sampleUsers[0]] }));
    useCreateUser.mockReturnValue({ mutate: vi.fn(), isPending: false });
    useUpdateUserByUserId.mockImplementation(() => ({
      mutate: vi.fn(),
      isPending: false,
    }));
    useDeleteUserByUserId.mockImplementation(() => ({
      mutate: deleteMutate,
      isPending: false,
    }));

    const confirmSpy = vi.spyOn(window, "confirm").mockReturnValue(true);
    renderUsers();
    fireEvent.click(screen.getByRole("button", { name: /delete/i }));
    expect(confirmSpy).toHaveBeenCalled();
    expect(deleteMutate).toHaveBeenCalledTimes(1);
    confirmSpy.mockRestore();
  });
});
