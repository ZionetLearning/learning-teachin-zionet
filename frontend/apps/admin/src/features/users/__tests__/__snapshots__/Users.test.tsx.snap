// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Users /> > matches snapshot (with users) 1`] = `
<DocumentFragment>
  <div
    class="root-0-2-1"
    data-testid="users-page"
  >
    <div
      class="formContainer-0-2-3 cardBase-0-2-2"
    >
      <h2
        class="sectionTitle-0-2-9"
      >
        pages.users.createUsers
      </h2>
      <form
        action="#"
        class="form-0-2-5"
      >
        <label
          class="label-0-2-6"
        >
          pages.users.email
          <input
            autocomplete="email"
            data-testid="users-create-email"
            name="email"
            placeholder="user@example.com"
            type="email"
            value=""
          />
          <span
            class="error-0-2-7"
          />
        </label>
        <label
          class="label-0-2-6"
        >
          pages.users.firstName
          <input
            autocomplete="given-name"
            data-testid="users-create-first-name"
            name="firstName"
            placeholder="John"
            type="text"
            value=""
          />
          <span
            class="error-0-2-7"
          />
        </label>
        <label
          class="label-0-2-6"
        >
          pages.users.lastName
          <input
            autocomplete="family-name"
            data-testid="users-create-last-name"
            name="lastName"
            placeholder="Doe"
            type="text"
            value=""
          />
          <span
            class="error-0-2-7"
          />
        </label>
        <label
          class="label-0-2-6"
        >
          pages.users.password
          <input
            autocomplete="new-password"
            data-testid="users-create-password"
            name="password"
            placeholder="******"
            type="password"
            value=""
          />
          <span
            class="error-0-2-7"
          />
        </label>
        <button
          class="submitButton-0-2-10"
          data-testid="users-create-submit"
          type="submit"
        >
          pages.users.createUser
        </button>
      </form>
    </div>
    <div
      class="listContainer-0-2-4 cardBase-0-2-2"
      data-testid="users-list-container"
    >
      <h2
        class="sectionTitle-0-2-9"
      >
        pages.users.users
      </h2>
      <ul
        class="list-0-2-8"
        data-testid="users-list"
      >
        <li
          class="listItem-0-2-11"
          data-testid="users-item-u1"
        >
          <div
            class="avatar-0-2-12"
          >
            A
          </div>
          <div
            class="info-0-2-13"
          >
            <span
              data-testid="users-email"
              title="alpha@example.com"
            >
              alpha@example.com
            </span>
            <span
              data-testid="users-first-name"
              title="Alice"
            >
              Alice
            </span>
            <span
              data-testid="users-last-name"
              title="Anderson"
            >
              Anderson
            </span>
          </div>
          <div
            class="actions-0-2-20"
          >
            <button
              class="updateButton-0-2-15 btnBase-0-2-14"
              data-testid="users-update-btn"
            >
              pages.users.update
            </button>
            <button
              class="deleteButton-0-2-16 btnBase-0-2-14"
              data-testid="users-delete-btn"
            >
              pages.users.delete
            </button>
          </div>
        </li>
        <li
          class="listItem-0-2-11"
          data-testid="users-item-u2"
        >
          <div
            class="avatar-0-2-12"
          >
            B
          </div>
          <div
            class="info-0-2-13"
          >
            <span
              data-testid="users-email"
              title="beta@example.com"
            >
              beta@example.com
            </span>
            <span
              data-testid="users-first-name"
              title="Bob"
            >
              Bob
            </span>
            <span
              data-testid="users-last-name"
              title="Baker"
            >
              Baker
            </span>
          </div>
          <div
            class="actions-0-2-20"
          >
            <button
              class="updateButton-0-2-15 btnBase-0-2-14"
              data-testid="users-update-btn"
            >
              pages.users.update
            </button>
            <button
              class="deleteButton-0-2-16 btnBase-0-2-14"
              data-testid="users-delete-btn"
            >
              pages.users.delete
            </button>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
